File: javascript/listas/linkedList.js
/*class Node {
    constructor(data, next = null){
        this.data = data;
        this.next = next;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
    }

    insert(data) {
        this.head = new Node(data, this.head);
    }


    displayList() {
        let current = this.head;
        let elements = []
        while(current){
            elements.push(current.data);
            current = current.next
        }
        console.log('[' + elements.join(', ') + ']')
    }
}

const linkedList = new LinkedList()

linkedList.insert(4);
linkedList.insert(45);
linkedList.displayList()*/






class Node {
    constructor(data){
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    constructor (){
        //Topo da lista 
        this.head = null;
        this.size = 0;
    }

    //Insere elemento ao final da lista
    push(data) {
        const newNode = new Node(data);
        if(!this.head) {
            // Se a lista está vazia, o novo nó se torna o primeiro nó (head)
            this.head = newNode;
        }else{
            //Se já existe algum elemento na lista, deve-se caminhar ao longo da sequência de nós
            //até chegar no último elemento.
            //último elemento: elemento em que o campo next é vazio
            //pointer = ponteiro que aponta para o próximo elemento da lista
            //pointer começa com o mesmo valor da cabeça
            let pointer = this.head;
            while(pointer.next){
                //Enquanto houver um next, significa que ainda há elementos na lista
                pointer = pointer.next;
            }
            // Adiciona o novo nó como próximo do último nó na lista
            pointer.next = newNode
        }
        this.size = this.size + 1;
    }

    //FUNÇÃO QUE REMOVE O ELEMENTO PASSADO NA FUNÇÃO
    remove(data) {
        if (!this.head) {
            return; // Lista vazia, nada a fazer
        }

        if (this.head.data === data) {
            this.head = this.head.next;
            this.size--;
            return;
        }

        let current = this.head;
        let previous = null;

        while (current && current.data !== data) {
            previous = current;
            current = current.next;
        }

        if (current) {
            previous.next = current.next;
            this.size--;
        }
    }

    searchValue(value) {
        let pointer = this.head;
        let position = 0;
        let found = false;
        while(pointer){
            if(pointer.data === value){
                found = true;
                console.log("Value " + value + " found at position " + position);
            }
            pointer = pointer.next;
            position ++;
        }
        if(!found){
            console.log("Element " + value + " not found");
        }
    }

    display() {
        let pointer = this.head;
        while(pointer){
            console.log(pointer.data);
            pointer = pointer.next;
        }
    }

}

const l1 = new LinkedList();
l1.push(3)
l1.push(7)
l1.push(4)
l1.push(22)
l1.remove(4)
l1.display();
l1.searchValue(33)
console.log("SIZE: " + l1.size)

File: javascript/listas/linkedList.c
/*#include <stdio.h>

typedef struct Node {
    int n;
    struct Node* next;
}Node;

int main () {
    Node n1, element2;

    n1.n = 13;
    n1.next;
    n1.next = &element2;
    element2.n = 3;
    element2.next = NULL;

    printf("First element: %d \n", n1.n);
    printf("Second element: %d \n", n1.next->n);
}*/

#include <stdio.h>
#include <stdlib.h>
typedef struct Node {
    int n;
    struct Node* next;
}Node;

int main () {
    Node n1;

    //Atribui valor 13 para o campo n do primeiro nó da lista
    n1.n = 13;
    //Malloc: para alocar memória para o próximo nó da lista / sizeof: tamanho necessário para armazenar um nó
    n1.next = malloc(sizeof(Node));
    //É atribuido o valor 4 para o campo n do próximo nó da lista
    n1.next->n = 4;
    //O campo next do próxmo nó da lista contém valor NULL, o qual indica que é este é o último nó
    n1.next->next = NULL;

    printf("First element: %d \n", n1.n);
    printf("Second element: %d \n", n1.next->n);
    //free() libera a memória alocada para o próximo nó da lista encadeada, apontado pelo ponteiro n1.next
    free(n1.next);
    /*Isso significa que a lista ligada possui dois campos: o que contém o elemento (n) e o outro que 
    contém o ponteiro que aponta para o próximo nó da lista*/
}
